lwt/util.ml

```diff
@@ -0,0 +1,35 @@
+open Xen_api
+open Lwt
+
+let sockaddr_of_uri uri =
+  let use_ssl =
+    match Uri.scheme uri with
+    | Some "https" ->
+      true
+    | Some "http" ->
+      false
+    | x ->
+      failwith (Printf.sprintf "Unsupported URI scheme: %s" (match x with None -> "None" | Some x -> x))
+  in
+  let port =
+    match Uri.port uri with
+    | None ->
+      (if use_ssl then 443 else 80)
+    | Some port ->
+      port
+  in
+  let host =
+    match Uri.host uri with
+    | None ->
+      fail (Failed_to_resolve_hostname "")
+    | Some host ->
+      Scanf.ksscanf host (fun _ _ -> host) "[%s@]" Fun.id
+  in
+  Lwt.catch (fun () ->
+    Lwt_unix.getaddrinfo host (string_of_int port) [] >|= function
+      | [] ->
+        return None
+      | addrinfo :: _ ->
+        return Some addrinfo.Unix.ai_addr)
```

Comment:
Wouldn't this work as well?
```suggestion
    Lwt_unix.getaddrinfo host (string_of_int port) [] >>= function
      | [] ->
        None
      | addrinfo :: _ ->
        Some addrinfo.Unix.ai_addr
```

Reply:
Yes, I'm not very familiar with Lwt stuff so i tried things to see if it solved my build error ^^'

Reply:
I believe what you need is `>|= function [] -> None | addrinfo :: _ -> Some addrinfo.Unix.ai_addr`

It should make a bit more sense if you see the signatures below. Basically `>|=` lets you use lwt values with normal functions

```
val (>>=) : 'a Lwt.t -> ('a -> 'b Lwt.t) -> 'b Lwt.t
val (>|=) : 'a Lwt.t -> ('a -> 'b) -> 'b Lwt.t
val return : 'a -> 'a t
```

Reply:
No worries, it's tough to get around the model of lwt with the monadic interface on top. In the end is promises and callback, just like in javascript or python.

I think there were tutorials about the principles in Lwt, I can look for the if you're interested. I had to [play around](https://gitlab.com/unduthegun/adventofcode2019/-/blob/master/00/intcode.ml) to actually understand how it works.